jaxopt/_src/gauss_newton.py

```diff
@@ -0,0 +1,165 @@
+      If True, the method ``run`` cannot be jitted.
+    implicit_diff_solve: the linear system solver to use.
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+    jit: whether to JIT-compile the bisection loop (default: "auto").
+    unroll: whether to unroll the bisection loop (default: "auto").
+  """
+  residual_fun: Callable
+  maxiter: int = 30
+  tol: float = 1e-5
+  check_bracket: bool = True
+  verbose: bool = False
+  implicit_diff_solve: Optional[Callable] = None
+  has_aux: bool = False
+  jit: base.AutoOrBoolean = "auto"
+  unroll: base.AutoOrBoolean = "auto"
+
+  def init_state(self,
+                 init_params: Optional[Any] = None,
+                 *args,
+                 **kwargs) -> GaussNewtonState:
+    """Initialize the solver state.
+
+    Args:
+      init_params: pytree containing the initial parameters.
+      *args: additional positional arguments to be passed to ``fun``.
+      **kwargs: additional keyword arguments to be passed to ``fun``.
+    Returns:
+      state
+    """
+
+    return GaussNewtonState(iter_num=jnp.asarray(0),
+                            error=jnp.asarray(jnp.inf),
+                            residual=jnp.asarray(jnp.inf),
+                            delta=jnp.asarray(jnp.inf),
+                            gradient=jnp.asarray(jnp.inf),
+                            jtj=jnp.asarray(jnp.inf))
+
+  def update(self,
+             params,
+             state: NamedTuple,
+             *args,
+             **kwargs) -> base.OptStep:
+    """Performs one iteration of the least-squares solver.
+
+    Args:
+      params: pytree containing the parameters.
+      state: named tuple containing the solver state.
+    Returns:
+      (params, state)
+    """
+    if params is None:
+      params = jnp.asarray(params, float)
+    else:
+      params = jnp.asarray(params, float)
+
+    residual, aux = self._fun_with_aux(params, *args, **kwargs)
+
+    matvec = lambda v: self._jtj(params, v, *args, **kwargs)
+    gradient = self._gradient(params, residual, *args, **kwargs)
+
+    delta_params = linear_solve.solve_cg(matvec, gradient)
+    params -= delta_params
+
+    state = GaussNewtonState(iter_num=state.iter_num + 1,
+                             error=jnp.linalg.norm(delta_params),
+                             residual=residual,
+                             delta=delta_params,
+                             gradient=gradient,
+                             aux=aux)
+
+    return base.OptStep(params=params, state=state)
+
+  def run(self,
+          init_params: Optional[Any] = None,
+          *args,
+          **kwargs) -> base.OptStep:
+    # We override run in order to set init_params=None by default.
+    return super().run(init_params, *args, **kwargs)
+
+  def __post_init__(self):
+    # We need this definition in the base solver run function
+    self.optimality_fun = self.residual_fun
```

Comment:
`optimality_fun` should contain the gradient of `0.5 ||residual_fun(params)||^2`, i.e. `optimality_fun(params) = J.T residual_fun(params)`. So if I'm not wrong it should be:
```python
def optimality_fun(params, *args, **kwargs):
  residual = self.residual_fun(params, *args, **kwargs)
  return self._gradient(self, params, residual, *args, **kwargs)
```

Reply:
Yes, that's a great point. This is essentially the RHS of the main equation we are solving to advance the params in each step.